driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -258,6 +329,80 @@ private Integer executeBatch(
+        BatchEncoder batchEncoder = new BatchEncoder();
+
+        AsyncCallbackSupplier<ExhaustiveClientBulkWriteCommandOkResponse> retryingBatchExecutor = decorateWriteWithRetriesAsync(
+                retryState, operationContext,
+                // Each batch re-selects a server and re-checks out a connection because this is simpler,
+                // and it is allowed by https://jira.mongodb.org/browse/DRIVERS-2502.
+                // If connection pinning is required, {@code binding} handles that,
+                // and `ClientSession`, `TransactionContext` are aware of that.
+                funcCallback -> withAsyncSourceAndConnection(binding::getWriteConnectionSource, true, funcCallback,
+                        (connectionSource, connection, resultCallback) -> {
+                            ConnectionDescription connectionDescription = connection.getDescription();
+                            boolean effectiveRetryWrites = isRetryableWrite(
+                                    retryWritesSetting, effectiveWriteConcern, connectionDescription, sessionContext);
+                            retryState.breakAndThrowIfRetryAnd(() -> !effectiveRetryWrites);
+                            resultAccumulator.onNewServerAddress(connectionDescription.getServerAddress());
+                            retryState.attach(AttachmentKeys.maxWireVersion(), connectionDescription.getMaxWireVersion(),
+                                            true)
+                                    .attach(AttachmentKeys.commandDescriptionSupplier(), () -> BULK_WRITE_COMMAND_NAME,
+                                            false);
+                            ClientBulkWriteCommand bulkWriteCommand = createBulkWriteCommand(
+                                    retryState, effectiveRetryWrites, effectiveWriteConcern, sessionContext,
+                                    unexecutedModels, batchEncoder,
+                                    () -> retryState.attach(AttachmentKeys.retryableCommandFlag(), true, true));
+                            executeBulkWriteCommandAndExhaustOkResponseAsync(
+                                    retryState, connectionSource, connection, bulkWriteCommand, effectiveWriteConcern,
+                                    operationContext,
+                                    resultCallback);
+                        }));
+
+        beginAsync()
+                .<ExhaustiveClientBulkWriteCommandOkResponse>thenSupply(callback -> {
+                    retryingBatchExecutor.get(callback);
+                })
+                .<Integer>thenApply((bulkWriteCommandOkResponse, callback) -> {
+                    callback.complete(resultAccumulator.onBulkWriteCommandOkResponseOrNoResponse(
+                            batchStartModelIndex, bulkWriteCommandOkResponse, batchEncoder.intoEncodedBatchInfo()));
+                }).onErrorIf(throwable -> true, (t, callback) -> {
```

Comment:
`throwable -> true` seems suspicious. This predicate is supposed to check the type of the exception, similarly how we do `catch(SpecifieExceptionType ...)`. Is it impossible to use the method here the way it is intended to be used?